crates/mun_gc/src/mark_sweep.rs

```diff
@@ -179,6 +173,61 @@ enum Color {
     Black,
 }
 
+/// Container for the `ObjectInfo` pool, and the map of `GcPtr`s to corresponding
+/// `ObjectInfo`.
+// Note: Right now, this roughly emulates the `slab` crate for the slab
+// allocator, which is thread safe because we wrap it in a `RwLock`. For future
+// perf gains, it might be worth investigating whether it's possible to get rid
+// of the lock and instead use a lockless allocator similar to `sharded-slab`.
+#[derive(Debug)]
+struct MarkSweepObjects<T: Type + Clone> {
+    /// A pinned slab allocator holding `ObjectInfo`.
+    ///
+    /// This is pinned to prevent accidental moves of individual elements, which
+    /// could cause `GcPtr`s to become invalid.
+    pub pool: Pool<ObjectInfo<T>>,
+
+    /// This maps each `GcPtr` to the index of its corresponding `ObjectInfo` in
+    /// the slab allocator.
+    pub info: HashMap<GcPtr, usize>,
+}
+
+impl<T: Type + Clone> MarkSweepObjects<T> {
+    pub fn new() -> Self {
+        MarkSweepObjects {
+            pool: Pool::new(),
+            info: HashMap::new(),
+        }
+    }
+
+    pub fn sweep(&mut self, observer: &impl Observer<Event = Event>, stats: &mut Stats) -> bool {
```

Comment:
I feel the code for this should reside somewhere in the `MarkSweep` object, where the tracing is also happening.